File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {   //1 1 2 3 4 4 5 6
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(ListNode curr : lists){
            while(curr != null){
                pq.offer(curr.val);
                curr = curr.next;
            }
        }
        ListNode res = new ListNode(0);
        ListNode head = res;
        while(!pq.isEmpty()){
            res.next = new ListNode(pq.poll());
            res = res.next;
        }
        return head.next;
        // 0->1 -> 2-> 2 -> 3 ->3 -> 4
    }
}

File: LeetCode/Hard/0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int k = lists.size();
        
        ListNode dummy(0);
        ListNode* head = &dummy;

        while(true) {
            int idx = -1;
            for (int i = 0; i < k; i++) {
                if (lists[i] == NULL) continue;

                if (idx == -1 || lists[i]->val < lists[idx]->val) {
                    idx = i;
                }
            }

            if (idx == -1) break;
            head->next = lists[idx];
            head = head->next;
            lists[idx] = lists[idx]->next;
        }

        return dummy.next;
    }
};